File: Leetcode/240125/Python/FindEventualSafeNode.py
# Problem Link: https://leetcode.com/problems/find-eventual-safe-states/description/?envType=daily-question&envId=2025-01-24
# Solution Link: 
# (i) BFS:  https://takeuforward.org/data-structure/find-eventual-safe-states-bfs-topological-sort-g-25/ 
# (ii) DFS: https://takeuforward.org/data-structure/find-eventual-safe-states-dfs-g-20/

class Solution(object):
    def eventualSafeNodes(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[int]
        """
        rows = len(graph)
        visited = [False]*rows
        pathVisited = [False]*rows

        res = []

        for i in range(len(graph)):
            if visited[i] == False:
               self.checkCycleDetection(i, graph, res, visited, pathVisited)
        res.sort()
        return res 
    def checkCycleDetection(self,node, graph, res, visited, pathVisited):
        visited[node] = True
        pathVisited[node] = True

        for i in range(len(graph[node])):
            currNode = graph[node][i]
            if visited[currNode] == False:
                if self.checkCycleDetection(currNode,graph, res, visited, pathVisited):
                    return True
            elif pathVisited[currNode] == True:
                   return True
        
        res.append(node)
        pathVisited[node] = False

        return False

File: Leetcode/240125/Java/FindEventualSafeNode.java
// Problem Link: https://leetcode.com/problems/find-eventual-safe-states/description/?envType=daily-question&envId=2025-01-24
// Solution Link: 
// (i) BFS:  https://takeuforward.org/data-structure/find-eventual-safe-states-bfs-topological-sort-g-25/ 
// (ii) DFS: https://takeuforward.org/data-structure/find-eventual-safe-states-dfs-g-20/

class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        
        boolean[] visited = new boolean[graph.length];
        boolean[] pathVisited = new boolean[graph.length];

        List<Integer> res = new ArrayList<>();

        for(int i=0; i<graph.length; i++){

            if(visited[i] == false){
              checkCycleDetection(i, graph, visited, pathVisited, res);
            }
        }

        Collections.sort(res);
        return res;
    }

    private boolean checkCycleDetection(int node, int[][] graph, boolean[] visited, 
    boolean[] pathVisited, List<Integer> res){

        visited[node] = true;
        pathVisited[node] = true;

        for(int i=0; i<graph[node].length; i++){

            int currNode = graph[node][i];

            if(visited[currNode] == false){

                if(checkCycleDetection(currNode, graph, visited, pathVisited, res)){
                    return true;
                }
            }
            else if(pathVisited[currNode]){
                return true;
            }
        }

        // If any node reach till here, it means this is first time, add it to result as safe node
        res.add(node);
        // Back track, make pathVisited as false
        pathVisited[node] = false;

        return false;


    }
}
